File: ios/PowderTracker/PowderTracker/Models/Event.swift
import Foundation
import SwiftUI

// MARK: - Urgency Level for Last Minute Events

enum UrgencyLevel {
    case departed   // Already left
    case critical   // < 1 hour
    case soon       // 1-3 hours
    case later      // > 3 hours
    case none       // No departure time set

    var color: Color {
        switch self {
        case .departed: return .gray
        case .critical: return .red
        case .soon: return .orange
        case .later: return .green
        case .none: return .secondary
        }
    }

    var label: String {
        switch self {
        case .departed: return "Departed"
        case .critical: return "Leaving Soon!"
        case .soon: return "Coming Up"
        case .later: return "Plenty of Time"
        case .none: return ""
        }
    }
}

// MARK: - Event Models

struct Event: Codable, Identifiable {
    let id: String
    let creatorId: String
    let mountainId: String
    let mountainName: String?
    let title: String
    let notes: String?
    let eventDate: String
    let departureTime: String?
    let departureLocation: String?
    let skillLevel: SkillLevel?
    let carpoolAvailable: Bool
    let carpoolSeats: Int?
    let status: EventStatus
    let createdAt: String
    let updatedAt: String
    let attendeeCount: Int
    let goingCount: Int
    let maybeCount: Int
    let creator: EventUser
    let userRSVPStatus: RSVPStatus?
    let isCreator: Bool?

    enum CodingKeys: String, CodingKey {
        case id
        case creatorId
        case mountainId
        case mountainName
        case title
        case notes
        case eventDate
        case departureTime
        case departureLocation
        case skillLevel
        case carpoolAvailable
        case carpoolSeats
        case status
        case createdAt
        case updatedAt
        case attendeeCount
        case goingCount
        case maybeCount
        case creator
        case userRSVPStatus
        case isCreator
    }

    // Static formatters to avoid expensive instantiation in computed properties
    private static let dateParser: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter
    }()

    private static let eventDateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "EEE, MMM d"
        return formatter
    }()

    private static let dateTimeParser: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd HH:mm"
        return formatter
    }()

    var formattedDate: String {
        guard let date = Self.dateParser.date(from: eventDate) else { return eventDate }
        return Self.eventDateFormatter.string(from: date)
    }

    var formattedTime: String? {
        guard let time = departureTime else { return nil }
        let components = time.split(separator: ":")
        guard components.count >= 2,
              let hour = Int(components[0]) else { return time }

        let h12 = hour % 12 == 0 ? 12 : hour % 12
        let ampm = hour >= 12 ? "PM" : "AM"
        return "\(h12):\(components[1]) \(ampm)"
    }

    // MARK: - Last Minute Crew Properties

    /// Whether this event is happening today
    var isToday: Bool {
        eventDate == Self.dateParser.string(from: Date())
    }

    /// Calculates time until departure in seconds
    var timeUntilDeparture: TimeInterval? {
        guard let time = departureTime else { return nil }

        guard let departureDate = Self.dateTimeParser.date(from: "\(eventDate) \(time)") else {
            return nil
        }

        return departureDate.timeIntervalSince(Date())
    }

    /// Urgency level based on time until departure
    var urgencyLevel: UrgencyLevel {
        guard let remaining = timeUntilDeparture else { return .none }
        switch remaining {
        case ..<0: return .departed
        case ..<3600: return .critical      // < 1 hour
        case ..<10800: return .soon         // 1-3 hours
        default: return .later              // > 3 hours
        }
    }

    /// Formatted countdown string (e.g., "2h 15m")
    var countdownText: String? {
        guard let remaining = timeUntilDeparture, remaining > 0 else { return nil }

        let hours = Int(remaining) / 3600
        let minutes = (Int(remaining) % 3600) / 60

        if hours > 0 {
            return "\(hours)h \(minutes)m"
        } else if minutes > 0 {
            return "\(minutes)m"
        } else {
            return "Now!"
        }
    }
}

struct EventWithDetails: Codable, Identifiable {
    let id: String
    let creatorId: String
    let mountainId: String
    let mountainName: String?
    let title: String
    let notes: String?
    let eventDate: String
    let departureTime: String?
    let departureLocation: String?
    let skillLevel: SkillLevel?
    let carpoolAvailable: Bool
    let carpoolSeats: Int?
    let status: EventStatus
    let createdAt: String
    let updatedAt: String
    let attendeeCount: Int
    let goingCount: Int
    let maybeCount: Int
    let commentCount: Int?
    let photoCount: Int?
    let creator: EventUser
    let userRSVPStatus: RSVPStatus?
    let isCreator: Bool?
    let attendees: [EventAttendee]
    let conditions: EventConditions?
    let inviteToken: String?

    enum CodingKeys: String, CodingKey {
        case id
        case creatorId
        case mountainId
        case mountainName
        case title
        case notes
        case eventDate
        case departureTime
        case departureLocation
        case skillLevel
        case carpoolAvailable
        case carpoolSeats
        case status
        case createdAt
        case updatedAt
        case attendeeCount
        case goingCount
        case maybeCount
        case commentCount
        case photoCount
        case creator
        case userRSVPStatus
        case isCreator
        case attendees
        case conditions
        case inviteToken
    }

    // Static formatters to avoid expensive instantiation in computed properties
    private static let dateParser: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter
    }()

    private static let fullDateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "EEEE, MMMM d, yyyy"
        return formatter
    }()

    var formattedDate: String {
        guard let date = Self.dateParser.date(from: eventDate) else { return eventDate }
        return Self.fullDateFormatter.string(from: date)
    }

    var formattedTime: String? {
        guard let time = departureTime else { return nil }
        let components = time.split(separator: ":")
        guard components.count >= 2,
              let hour = Int(components[0]) else { return time }

        let h12 = hour % 12 == 0 ? 12 : hour % 12
        let ampm = hour >= 12 ? "PM" : "AM"
        return "\(h12):\(components[1]) \(ampm)"
    }
}

struct EventUser: Codable {
    let id: String
    let username: String
    let displayName: String?
    let avatarUrl: String?

    enum CodingKeys: String, CodingKey {
        case id
        case username
        case displayName = "display_name"
        case avatarUrl = "avatar_url"
    }

    var displayNameOrUsername: String {
        displayName ?? username
    }
}

struct EventAttendee: Codable, Identifiable {
    let id: String
    let userId: String
    let status: RSVPStatus
    let isDriver: Bool
    let needsRide: Bool
    let pickupLocation: String?
    let respondedAt: String?
    let user: EventUser

    enum CodingKeys: String, CodingKey {
        case id
        case userId
        case status
        case isDriver
        case needsRide
        case pickupLocation
        case respondedAt
        case user
    }
}

struct EventConditions: Codable {
    let temperature: Double?
    let snowfall24h: Double?
    let snowDepth: Double?
    let powderScore: Double?
    let forecast: EventForecast?
}

struct EventForecast: Codable {
    let high: Double
    let low: Double
    let snowfall: Double
    let conditions: String
}

// MARK: - Enums

enum EventStatus: String, Codable {
    case active
    case cancelled
    case completed
}

enum SkillLevel: String, Codable, CaseIterable {
    case beginner
    case intermediate
    case advanced
    case expert
    case all

    var displayName: String {
        switch self {
        case .beginner: return "Beginner"
        case .intermediate: return "Intermediate"
        case .advanced: return "Advanced"
        case .expert: return "Expert"
        case .all: return "All Levels"
        }
    }
}

enum RSVPStatus: String, Codable {
    case invited
    case going
    case maybe
    case declined

    var displayName: String {
        switch self {
        case .invited: return "Invited"
        case .going: return "Going"
        case .maybe: return "Maybe"
        case .declined: return "Not Going"
        }
    }

    var color: String {
        switch self {
        case .going: return "green"
        case .maybe: return "yellow"
        case .invited, .declined: return "gray"
        }
    }
}

// MARK: - Response Types

struct EventsListResponse: Codable {
    let events: [Event]
    let pagination: EventPagination
}

struct EventPagination: Codable {
    let total: Int
    let limit: Int
    let offset: Int
    let hasMore: Bool
}

struct EventResponse: Codable {
    let event: EventWithDetails
}

struct CreateEventResponse: Codable {
    let event: Event
    let inviteToken: String
    let inviteUrl: String
}

struct RSVPResponse: Codable {
    let attendee: EventAttendee
    let event: RSVPEventUpdate
}

struct RSVPEventUpdate: Codable {
    let id: String
    let goingCount: Int
    let maybeCount: Int
    let attendeeCount: Int
}

struct InviteResponse: Codable {
    let invite: InviteInfo
}

struct InviteInfo: Codable {
    let event: Event
    let conditions: EventConditions?
    let isValid: Bool
    let isExpired: Bool
    let requiresAuth: Bool
}

// MARK: - Request Types

struct CreateEventRequest: Encodable {
    let mountainId: String
    let title: String
    let notes: String?
    let eventDate: String
    let departureTime: String?
    let departureLocation: String?
    let skillLevel: String?
    let carpoolAvailable: Bool?
    let carpoolSeats: Int?
}

struct RSVPRequest: Encodable {
    let status: String
    let isDriver: Bool?
    let needsRide: Bool?
    let pickupLocation: String?
}

File: src/types/event.ts
/**
 * TypeScript types for ski events social feature
 */

// ============================================
// Enums / Literal Types
// ============================================

export type EventStatus = 'active' | 'cancelled' | 'completed';

export type SkillLevel = 'beginner' | 'intermediate' | 'advanced' | 'expert' | 'all';

export type RSVPStatus = 'invited' | 'going' | 'maybe' | 'declined' | 'waitlist';

// ============================================
// Database Row Types
// ============================================

export interface EventRow {
  id: string;
  user_id: string;
  mountain_id: string;
  title: string;
  notes: string | null;
  event_date: string; // DATE as ISO string
  departure_time: string | null; // TIME as string (HH:MM:SS)
  departure_location: string | null;
  skill_level: SkillLevel | null;
  carpool_available: boolean;
  carpool_seats: number | null;
  max_attendees: number | null;
  status: EventStatus;
  created_at: string;
  updated_at: string;
  attendee_count: number;
  going_count: number;
  maybe_count: number;
  waitlist_count: number;
}

export interface EventAttendeeRow {
  id: string;
  event_id: string;
  user_id: string;
  status: RSVPStatus;
  is_driver: boolean;
  needs_ride: boolean;
  pickup_location: string | null;
  passengers_count: number;
  waitlist_position: number | null;
  responded_at: string | null;
  invited_at: string;
  created_at: string;
  updated_at: string;
}

export interface EventInviteTokenRow {
  id: string;
  event_id: string;
  token: string;
  created_by: string;
  uses_count: number;
  max_uses: number | null;
  expires_at: string | null;
  created_at: string;
}

// ============================================
// Insert Types (for creating new records)
// ============================================

export interface EventInsert {
  user_id: string;
  mountain_id: string;
  title: string;
  notes?: string | null;
  event_date: string;
  departure_time?: string | null;
  departure_location?: string | null;
  skill_level?: SkillLevel | null;
  carpool_available?: boolean;
  carpool_seats?: number | null;
  max_attendees?: number | null;
}

export interface EventAttendeeInsert {
  event_id: string;
  user_id: string;
  status?: RSVPStatus;
  is_driver?: boolean;
  needs_ride?: boolean;
  pickup_location?: string | null;
  passengers_count?: number;
}

export interface EventInviteTokenInsert {
  event_id: string;
  token: string;
  created_by: string;
  max_uses?: number | null;
  expires_at?: string | null;
}

// ============================================
// Update Types (for modifying records)
// ============================================

export interface EventUpdate {
  title?: string;
  notes?: string | null;
  event_date?: string;
  departure_time?: string | null;
  departure_location?: string | null;
  skill_level?: SkillLevel | null;
  carpool_available?: boolean;
  carpool_seats?: number | null;
  max_attendees?: number | null;
  status?: EventStatus;
}

export interface EventAttendeeUpdate {
  status?: RSVPStatus;
  is_driver?: boolean;
  needs_ride?: boolean;
  pickup_location?: string | null;
  passengers_count?: number;
  responded_at?: string;
}

// ============================================
// API Request Types
// ============================================

export interface CreateEventRequest {
  mountainId: string;
  title: string;
  notes?: string;
  eventDate: string; // ISO date string (YYYY-MM-DD)
  departureTime?: string; // HH:MM format
  departureLocation?: string;
  skillLevel?: SkillLevel;
  carpoolAvailable?: boolean;
  carpoolSeats?: number;
  maxAttendees?: number; // Optional capacity limit (1-1000)
}

export interface UpdateEventRequest {
  title?: string;
  notes?: string | null;
  eventDate?: string;
  departureTime?: string | null;
  departureLocation?: string | null;
  skillLevel?: SkillLevel | null;
  carpoolAvailable?: boolean;
  carpoolSeats?: number | null;
  maxAttendees?: number | null; // Optional capacity limit (1-1000, null = unlimited)
}

export interface RSVPRequest {
  status: 'going' | 'maybe' | 'declined';
  isDriver?: boolean;
  needsRide?: boolean;
  pickupLocation?: string;
}

// ============================================
// API Response Types
// ============================================

export interface EventUser {
  id: string;
  username: string;
  display_name: string | null;
  avatar_url: string | null;
}

export interface EventAttendee {
  id: string;
  userId: string;
  status: RSVPStatus;
  isDriver: boolean;
  needsRide: boolean;
  pickupLocation: string | null;
  waitlistPosition: number | null;
  respondedAt: string | null;
  user: EventUser;
}

export interface EventConditions {
  temperature?: number;
  snowfall24h?: number;
  snowDepth?: number;
  powderScore?: number;
  forecast?: {
    high: number;
    low: number;
    snowfall: number;
    conditions: string;
  };
}

export interface Event {
  id: string;
  creatorId: string;
  mountainId: string;
  mountainName?: string;
  title: string;
  notes: string | null;
  eventDate: string;
  departureTime: string | null;
  departureLocation: string | null;
  skillLevel: SkillLevel | null;
  carpoolAvailable: boolean;
  carpoolSeats: number | null;
  maxAttendees: number | null;
  status: EventStatus;
  createdAt: string;
  updatedAt: string;
  attendeeCount: number;
  goingCount: number;
  maybeCount: number;
  waitlistCount: number;
  commentCount?: number;
  photoCount?: number;
  creator: EventUser;
  userRSVPStatus?: RSVPStatus | null;
  isCreator?: boolean;
}

export interface EventWithDetails extends Event {
  attendees: EventAttendee[];
  conditions?: EventConditions;
  inviteToken?: string;
}

export interface EventsListResponse {
  events: Event[];
  pagination: {
    total: number;
    limit: number;
    offset: number;
    hasMore: boolean;
  };
}

export interface EventResponse {
  event: EventWithDetails;
}

export interface CreateEventResponse {
  event: Event;
  inviteToken: string;
  inviteUrl: string;
}

export interface RSVPResponse {
  attendee: EventAttendee;
  event: {
    id: string;
    goingCount: number;
    maybeCount: number;
    attendeeCount: number;
    waitlistCount?: number;
    maxAttendees?: number | null;
  };
  message?: string;
  wasWaitlisted?: boolean;
}

// ============================================
// Invite Link Types
// ============================================

export interface InviteInfo {
  event: Event;
  conditions?: EventConditions;
  isValid: boolean;
  isExpired: boolean;
  requiresAuth: boolean;
}

export interface InviteResponse {
  invite: InviteInfo;
}

// ============================================
// Push Notification Types
// ============================================

export interface EventReminderPayload {
  eventId: string;
  eventTitle: string;
  mountainId: string;
  mountainName: string;
  eventDate: string;
  departureTime: string | null;
  goingCount: number;
  conditions?: EventConditions;
}

// ============================================
// Date Polling Types
// ============================================

export type DateVoteChoice = 'available' | 'maybe' | 'unavailable';

export type DatePollStatus = 'open' | 'closed';

export interface DatePollOption {
  id: string;
  proposedDate: string; // YYYY-MM-DD
  proposedBy: string;
  votes: DatePollVote[];
  availableCount: number;
  maybeCount: number;
  unavailableCount: number;
}

export interface DatePollVote {
  userId: string;
  vote: DateVoteChoice;
  user?: EventUser;
}

export interface DatePoll {
  id: string;
  eventId: string;
  status: DatePollStatus;
  createdAt: string;
  closedAt: string | null;
  options: DatePollOption[];
}

export interface DatePollResponse {
  poll: DatePoll;
}

export interface CreateDatePollRequest {
  dates: string[]; // Array of YYYY-MM-DD strings (2-5 dates)
}

export interface CastDateVoteRequest {
  optionId: string;
  vote: DateVoteChoice;
}

export interface ResolveDatePollRequest {
  optionId: string; // The winning date option
}

// ============================================
// Query Parameters Types
// ============================================

export interface EventsQueryParams {
  mountainId?: string;
  status?: EventStatus;
  upcoming?: boolean;
  createdByMe?: boolean;
  attendingOnly?: boolean;
  limit?: number;
  offset?: number;
}

// ============================================
// Comment Types
// ============================================

export interface EventComment {
  id: string;
  event_id: string;
  user_id: string;
  content: string;
  parent_id: string | null;
  created_at: string;
  updated_at: string;
  user: EventUser;
  replies?: EventComment[];
}

export interface CommentsResponse {
  comments: EventComment[];
  commentCount: number;
  gated: boolean;
  message?: string;
}

export interface CreateCommentRequest {
  content: string;
  parentId?: string;
}

export interface CreateCommentResponse {
  comment: EventComment;
}

// ============================================
// Activity Types
// ============================================

export type ActivityType =
  | 'rsvp_going'
  | 'rsvp_maybe'
  | 'rsvp_declined'
  | 'comment_posted'
  | 'milestone_reached'
  | 'event_created'
  | 'event_updated';

export interface ActivityMetadata {
  milestone?: number;
  label?: string;
  comment_id?: string;
  preview?: string;
  is_reply?: boolean;
  previous_status?: string;
}

export interface EventActivity {
  id: string;
  eventId: string;
  userId: string | null;
  activityType: ActivityType;
  metadata: ActivityMetadata;
  createdAt: string;
  user?: {
    id: string;
    username: string;
    displayName: string | null;
    avatarUrl: string | null;
  } | null;
}

export interface ActivityResponse {
  activities: EventActivity[];
  activityCount: number;
  gated: boolean;
  message?: string;
  pagination?: {
    limit: number;
    offset: number;
    hasMore: boolean;
  };
}
